// luis-apis/lib/operations/azureAccounts.js
"use strict";
/*
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const msRest = require("@azure/ms-rest-js");
const Mappers = require("../models/azureAccountsMappers");
const Parameters = require("../models/parameters");
/** Class representing a AzureAccounts. */
class AzureAccounts {
    /**
     * Create a AzureAccounts.
     * @param {LuisAuthoringContext} client Reference to the service client.
     */
    constructor(client) {
        this.client = client;
    }
    assignToApp(appId, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            options
        }, assignToAppOperationSpec, callback);
    }
    getAssigned(appId, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            options
        }, getAssignedOperationSpec, callback);
    }
    removeFromApp(appId, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            options
        }, removeFromAppOperationSpec, callback);
    }
    listUserLUISAccounts(options, callback) {
        return this.client.sendOperationRequest({
            options
        }, listUserLUISAccountsOperationSpec, callback);
    }
}
exports.AzureAccounts = AzureAccounts;
// Operation Specifications
const serializer = new msRest.Serializer(Mappers);
const assignToAppOperationSpec = {
    httpMethod: "POST",
    path: "apps/{appId}/azureaccounts",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId
    ],
    requestBody: {
        parameterPath: [
            "options",
            "azureAccountInfoObject"
        ],
        mapper: Mappers.AzureAccountInfoObject
    },
    responses: {
        201: {
            bodyMapper: Mappers.OperationStatus
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const getAssignedOperationSpec = {
    httpMethod: "GET",
    path: "apps/{appId}/azureaccounts",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId
    ],
    responses: {
        200: {
            bodyMapper: {
                serializedName: "parsedResponse",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "AzureAccountInfoObject"
                        }
                    }
                }
            }
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const removeFromAppOperationSpec = {
    httpMethod: "DELETE",
    path: "apps/{appId}/azureaccounts",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId
    ],
    requestBody: {
        parameterPath: [
            "options",
            "azureAccountInfoObject"
        ],
        mapper: Mappers.AzureAccountInfoObject
    },
    responses: {
        200: {
            bodyMapper: Mappers.OperationStatus
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const listUserLUISAccountsOperationSpec = {
    httpMethod: "GET",
    path: "azureaccounts",
    urlParameters: [
        Parameters.endpoint
    ],
    responses: {
        200: {
            bodyMapper: {
                serializedName: "parsedResponse",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "AzureAccountInfoObject"
                        }
                    }
                }
            }
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
