// luis-apis/lib/operations/examples.js
"use strict";
/*
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const msRest = require("@azure/ms-rest-js");
const Mappers = require("../models/examplesMappers");
const Parameters = require("../models/parameters");
/** Class representing a Examples. */
class Examples {
    /**
     * Create a Examples.
     * @param {LuisAuthoringContext} client Reference to the service client.
     */
    constructor(client) {
        this.client = client;
    }
    add(appId, versionId, exampleLabelObject, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            exampleLabelObject,
            options
        }, addOperationSpec, callback);
    }
    batch(appId, versionId, exampleLabelObjectArray, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            exampleLabelObjectArray,
            options
        }, batchOperationSpec, callback);
    }
    list(appId, versionId, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            options
        }, listOperationSpec, callback);
    }
    deleteMethod(appId, versionId, exampleId, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            exampleId,
            options
        }, deleteMethodOperationSpec, callback);
    }
}
exports.Examples = Examples;
// Operation Specifications
const serializer = new msRest.Serializer(Mappers);
const addOperationSpec = {
    httpMethod: "POST",
    path: "apps/{appId}/versions/{versionId}/example",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0
    ],
    requestBody: {
        parameterPath: "exampleLabelObject",
        mapper: Object.assign({}, Mappers.ExampleLabelObject, { required: true })
    },
    responses: {
        201: {
            bodyMapper: Mappers.LabelExampleResponse
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const batchOperationSpec = {
    httpMethod: "POST",
    path: "apps/{appId}/versions/{versionId}/examples",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0
    ],
    requestBody: {
        parameterPath: "exampleLabelObjectArray",
        mapper: {
            required: true,
            serializedName: "exampleLabelObjectArray",
            type: {
                name: "Sequence",
                element: {
                    type: {
                        name: "Composite",
                        className: "ExampleLabelObject"
                    }
                }
            }
        }
    },
    responses: {
        201: {
            bodyMapper: {
                serializedName: "parsedResponse",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "BatchLabelExample"
                        }
                    }
                }
            }
        },
        207: {
            bodyMapper: {
                serializedName: "parsedResponse",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "BatchLabelExample"
                        }
                    }
                }
            }
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const listOperationSpec = {
    httpMethod: "GET",
    path: "apps/{appId}/versions/{versionId}/examples",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0
    ],
    queryParameters: [
        Parameters.skip,
        Parameters.take
    ],
    responses: {
        200: {
            bodyMapper: {
                serializedName: "parsedResponse",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "LabeledUtterance"
                        }
                    }
                }
            }
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const deleteMethodOperationSpec = {
    httpMethod: "DELETE",
    path: "apps/{appId}/versions/{versionId}/examples/{exampleId}",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0,
        Parameters.exampleId
    ],
    responses: {
        200: {
            bodyMapper: Mappers.OperationStatus
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
