// luis-apis/lib/operations/pattern.js
"use strict";
/*
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const msRest = require("@azure/ms-rest-js");
const Mappers = require("../models/patternMappers");
const Parameters = require("../models/parameters");
/** Class representing a Pattern. */
class Pattern {
    /**
     * Create a Pattern.
     * @param {LuisAuthoringContext} client Reference to the service client.
     */
    constructor(client) {
        this.client = client;
    }
    addPattern(appId, versionId, pattern, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            pattern,
            options
        }, addPatternOperationSpec, callback);
    }
    listPatterns(appId, versionId, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            options
        }, listPatternsOperationSpec, callback);
    }
    updatePatterns(appId, versionId, patterns, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            patterns,
            options
        }, updatePatternsOperationSpec, callback);
    }
    batchAddPatterns(appId, versionId, patterns, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            patterns,
            options
        }, batchAddPatternsOperationSpec, callback);
    }
    deletePatterns(appId, versionId, patternIds, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            patternIds,
            options
        }, deletePatternsOperationSpec, callback);
    }
    updatePattern(appId, versionId, patternId, pattern, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            patternId,
            pattern,
            options
        }, updatePatternOperationSpec, callback);
    }
    deletePattern(appId, versionId, patternId, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            patternId,
            options
        }, deletePatternOperationSpec, callback);
    }
    listIntentPatterns(appId, versionId, intentId, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            intentId,
            options
        }, listIntentPatternsOperationSpec, callback);
    }
}
exports.Pattern = Pattern;
// Operation Specifications
const serializer = new msRest.Serializer(Mappers);
const addPatternOperationSpec = {
    httpMethod: "POST",
    path: "apps/{appId}/versions/{versionId}/patternrule",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0
    ],
    requestBody: {
        parameterPath: "pattern",
        mapper: Object.assign({}, Mappers.PatternRuleCreateObject, { required: true })
    },
    responses: {
        201: {
            bodyMapper: Mappers.PatternRuleInfo
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const listPatternsOperationSpec = {
    httpMethod: "GET",
    path: "apps/{appId}/versions/{versionId}/patternrules",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0
    ],
    queryParameters: [
        Parameters.skip,
        Parameters.take
    ],
    responses: {
        200: {
            bodyMapper: {
                serializedName: "parsedResponse",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "PatternRuleInfo"
                        }
                    }
                }
            }
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const updatePatternsOperationSpec = {
    httpMethod: "PUT",
    path: "apps/{appId}/versions/{versionId}/patternrules",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0
    ],
    requestBody: {
        parameterPath: "patterns",
        mapper: {
            required: true,
            serializedName: "patterns",
            type: {
                name: "Sequence",
                element: {
                    type: {
                        name: "Composite",
                        className: "PatternRuleUpdateObject"
                    }
                }
            }
        }
    },
    responses: {
        200: {
            bodyMapper: {
                serializedName: "parsedResponse",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "PatternRuleInfo"
                        }
                    }
                }
            }
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const batchAddPatternsOperationSpec = {
    httpMethod: "POST",
    path: "apps/{appId}/versions/{versionId}/patternrules",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0
    ],
    requestBody: {
        parameterPath: "patterns",
        mapper: {
            required: true,
            serializedName: "patterns",
            type: {
                name: "Sequence",
                element: {
                    type: {
                        name: "Composite",
                        className: "PatternRuleCreateObject"
                    }
                }
            }
        }
    },
    responses: {
        201: {
            bodyMapper: {
                serializedName: "parsedResponse",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "PatternRuleInfo"
                        }
                    }
                }
            }
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const deletePatternsOperationSpec = {
    httpMethod: "DELETE",
    path: "apps/{appId}/versions/{versionId}/patternrules",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0
    ],
    requestBody: {
        parameterPath: "patternIds",
        mapper: {
            required: true,
            serializedName: "patternIds",
            type: {
                name: "Sequence",
                element: {
                    type: {
                        name: "Uuid"
                    }
                }
            }
        }
    },
    responses: {
        200: {
            bodyMapper: Mappers.OperationStatus
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const updatePatternOperationSpec = {
    httpMethod: "PUT",
    path: "apps/{appId}/versions/{versionId}/patternrules/{patternId}",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0,
        Parameters.patternId1
    ],
    requestBody: {
        parameterPath: "pattern",
        mapper: Object.assign({}, Mappers.PatternRuleUpdateObject, { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.PatternRuleInfo
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const deletePatternOperationSpec = {
    httpMethod: "DELETE",
    path: "apps/{appId}/versions/{versionId}/patternrules/{patternId}",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0,
        Parameters.patternId1
    ],
    responses: {
        200: {
            bodyMapper: Mappers.OperationStatus
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const listIntentPatternsOperationSpec = {
    httpMethod: "GET",
    path: "apps/{appId}/versions/{versionId}/intents/{intentId}/patternrules",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0,
        Parameters.intentId
    ],
    queryParameters: [
        Parameters.skip,
        Parameters.take
    ],
    responses: {
        200: {
            bodyMapper: {
                serializedName: "parsedResponse",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "PatternRuleInfo"
                        }
                    }
                }
            }
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
