// luis-apis/lib/operations/settings.js
"use strict";
/*
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const msRest = require("@azure/ms-rest-js");
const Mappers = require("../models/settingsMappers");
const Parameters = require("../models/parameters");
/** Class representing a Settings. */
class Settings {
    /**
     * Create a Settings.
     * @param {LuisAuthoringContext} client Reference to the service client.
     */
    constructor(client) {
        this.client = client;
    }
    list(appId, versionId, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            options
        }, listOperationSpec, callback);
    }
    update(appId, versionId, listOfAppVersionSettingObject, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            listOfAppVersionSettingObject,
            options
        }, updateOperationSpec, callback);
    }
}
exports.Settings = Settings;
// Operation Specifications
const serializer = new msRest.Serializer(Mappers);
const listOperationSpec = {
    httpMethod: "GET",
    path: "apps/{appId}/versions/{versionId}/settings",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0
    ],
    responses: {
        200: {
            bodyMapper: {
                serializedName: "parsedResponse",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "AppVersionSettingObject"
                        }
                    }
                }
            }
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const updateOperationSpec = {
    httpMethod: "PUT",
    path: "apps/{appId}/versions/{versionId}/settings",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0
    ],
    requestBody: {
        parameterPath: "listOfAppVersionSettingObject",
        mapper: {
            required: true,
            serializedName: "listOfAppVersionSettingObject",
            type: {
                name: "Sequence",
                element: {
                    type: {
                        name: "Composite",
                        className: "AppVersionSettingObject"
                    }
                }
            }
        }
    },
    responses: {
        200: {
            bodyMapper: Mappers.OperationStatus
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
