// luis-apis/lib/operations/train.js
"use strict";
/*
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const msRest = require("@azure/ms-rest-js");
const Mappers = require("../models/trainMappers");
const Parameters = require("../models/parameters");
/** Class representing a Train. */
class Train {
    /**
     * Create a Train.
     * @param {LuisAuthoringContext} client Reference to the service client.
     */
    constructor(client) {
        this.client = client;
    }
    trainVersion(appId, versionId, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            options
        }, trainVersionOperationSpec, callback);
    }
    getStatus(appId, versionId, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            options
        }, getStatusOperationSpec, callback);
    }
}
exports.Train = Train;
// Operation Specifications
const serializer = new msRest.Serializer(Mappers);
const trainVersionOperationSpec = {
    httpMethod: "POST",
    path: "apps/{appId}/versions/{versionId}/train",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0
    ],
    responses: {
        202: {
            bodyMapper: Mappers.EnqueueTrainingResponse
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const getStatusOperationSpec = {
    httpMethod: "GET",
    path: "apps/{appId}/versions/{versionId}/train",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0
    ],
    responses: {
        200: {
            bodyMapper: {
                serializedName: "parsedResponse",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "ModelTrainingInfo"
                        }
                    }
                }
            }
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
