// luis-apis/lib/operations/versions.js
"use strict";
/*
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const msRest = require("@azure/ms-rest-js");
const Mappers = require("../models/versionsMappers");
const Parameters = require("../models/parameters");
/** Class representing a Versions. */
class Versions {
    /**
     * Create a Versions.
     * @param {LuisAuthoringContext} client Reference to the service client.
     */
    constructor(client) {
        this.client = client;
    }
    clone(appId, versionId, versionCloneObject, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            versionCloneObject,
            options
        }, cloneOperationSpec, callback);
    }
    list(appId, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            options
        }, listOperationSpec, callback);
    }
    get(appId, versionId, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            options
        }, getOperationSpec, callback);
    }
    update(appId, versionId, versionUpdateObject, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            versionUpdateObject,
            options
        }, updateOperationSpec, callback);
    }
    deleteMethod(appId, versionId, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            options
        }, deleteMethodOperationSpec, callback);
    }
    exportMethod(appId, versionId, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            options
        }, exportMethodOperationSpec, callback);
    }
    importMethod(appId, luisApp, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            luisApp,
            options
        }, importMethodOperationSpec, callback);
    }
    deleteUnlabelledUtterance(appId, versionId, utterance, options, callback) {
        return this.client.sendOperationRequest({
            appId,
            versionId,
            utterance,
            options
        }, deleteUnlabelledUtteranceOperationSpec, callback);
    }
}
exports.Versions = Versions;
// Operation Specifications
const serializer = new msRest.Serializer(Mappers);
const cloneOperationSpec = {
    httpMethod: "POST",
    path: "apps/{appId}/versions/{versionId}/clone",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0
    ],
    requestBody: {
        parameterPath: "versionCloneObject",
        mapper: Object.assign({}, Mappers.TaskUpdateObject, { required: true })
    },
    responses: {
        201: {
            bodyMapper: {
                serializedName: "parsedResponse",
                type: {
                    name: "String"
                }
            }
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const listOperationSpec = {
    httpMethod: "GET",
    path: "apps/{appId}/versions",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId
    ],
    queryParameters: [
        Parameters.skip,
        Parameters.take
    ],
    responses: {
        200: {
            bodyMapper: {
                serializedName: "parsedResponse",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "VersionInfo"
                        }
                    }
                }
            }
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const getOperationSpec = {
    httpMethod: "GET",
    path: "apps/{appId}/versions/{versionId}/",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0
    ],
    responses: {
        200: {
            bodyMapper: Mappers.VersionInfo
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const updateOperationSpec = {
    httpMethod: "PUT",
    path: "apps/{appId}/versions/{versionId}/",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0
    ],
    requestBody: {
        parameterPath: "versionUpdateObject",
        mapper: Object.assign({}, Mappers.TaskUpdateObject, { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.OperationStatus
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const deleteMethodOperationSpec = {
    httpMethod: "DELETE",
    path: "apps/{appId}/versions/{versionId}/",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0
    ],
    responses: {
        200: {
            bodyMapper: Mappers.OperationStatus
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const exportMethodOperationSpec = {
    httpMethod: "GET",
    path: "apps/{appId}/versions/{versionId}/export",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0
    ],
    responses: {
        200: {
            bodyMapper: Mappers.LuisApp
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const importMethodOperationSpec = {
    httpMethod: "POST",
    path: "apps/{appId}/versions/import",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId
    ],
    queryParameters: [
        Parameters.versionId1
    ],
    requestBody: {
        parameterPath: "luisApp",
        mapper: Object.assign({}, Mappers.LuisApp, { required: true })
    },
    responses: {
        201: {
            bodyMapper: {
                serializedName: "parsedResponse",
                type: {
                    name: "String"
                }
            }
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const deleteUnlabelledUtteranceOperationSpec = {
    httpMethod: "DELETE",
    path: "apps/{appId}/versions/{versionId}/suggest",
    urlParameters: [
        Parameters.endpoint,
        Parameters.appId,
        Parameters.versionId0
    ],
    requestBody: {
        parameterPath: "utterance",
        mapper: {
            required: true,
            serializedName: "utterance",
            type: {
                name: "String"
            }
        }
    },
    responses: {
        200: {
            bodyMapper: Mappers.OperationStatus
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
