// luis-apis/lib/luisAuthoring.js
"use strict";
/*
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
const Models = require("./models");
exports.LuisAuthoringModels = Models;
const Mappers = require("./models/mappers");
exports.LuisAuthoringMappers = Mappers;
const operations = require("./operations");
const luisAuthoringContext_1 = require("./luisAuthoringContext");
exports.LuisAuthoringContext = luisAuthoringContext_1.LuisAuthoringContext;
class LuisAuthoring extends luisAuthoringContext_1.LuisAuthoringContext {
    /**
     * Initializes a new instance of the LuisAuthoring class.
     * @param endpoint Supported Cognitive Services endpoints (protocol and hostname, for example:
     * https://westus.api.cognitive.microsoft.com).
     * @param credentials Subscription credentials which uniquely identify client subscription.
     * @param [options] The parameter options
     */
    constructor(endpoint, credentials, options) {
        super(endpoint, credentials, options);
        this.features = new operations.Features(this);
        this.examples = new operations.Examples(this);
        this.model = new operations.Model(this);
        this.apps = new operations.Apps(this);
        this.versions = new operations.Versions(this);
        this.train = new operations.Train(this);
        this.permissions = new operations.Permissions(this);
        this.pattern = new operations.Pattern(this);
        this.settings = new operations.Settings(this);
        this.azureAccounts = new operations.AzureAccounts(this);
    }
}
exports.LuisAuthoring = LuisAuthoring;
__export(require("./operations"));
